libraries/AP_Scripting/examples/ahrs-source.lua

```diff
@@ -0,0 +1,177 @@
+local source_prev = 0               -- previous source, defaults to primary source
+local sw_source_prev = -1           -- previous source switch position
+local sw_auto_pos_prev = -1         -- previous auto source switch position
+local auto_switch = false           -- true when auto switching between sources is active
+local gps_usable_accuracy = 1.0     -- GPS is usable if speed accuracy is at or below this value
+local vote_counter_max = 20         -- when a vote counter reaches this number (i.e. 2sec) source may be switched
+local gps_vs_nongps_vote = 0        -- vote counter for GPS vs NonGPS (-20 = GPS, +20 = NonGPS)
+local extnav_vs_opticalflow_vote = 0 -- vote counter for extnav vs optical flow (-20 = extnav, +20 = opticalflow)
+
+local debug_counter = 0
+
+-- the main update function that uses the takeoff and velocity controllers to fly a rough square pattern
+function update()
+
+  -- check switches are configured
+  -- source selection from RCx_FUNCTION = 90 (EKF Source Select)
+  -- auto source from RCx_FUNCTION = 83 (ZigZag_Auto)
+  local rc_function_source = rc:find_channel_for_option(90)
+  local rc_function_auto = rc:find_channel_for_option(83)
+  if (rc_function_source == nil) or (rc_function_auto == nil) then
+    gcs:send_text(0, "ahrs-source.lua: RCx_FUNCTION=90 or 83 not set!")
+    return update, 1000
+  end
+
+  -- check rangefinder distance threshold has been set
+  local rangefinder_thresh_dist = param:get('SCR_USER1')  -- SCR_USER1 holds rangefinder threshold
+  if (rangefinder_thresh_dist == nil) or (rangefinder_thresh_dist <= 0) then
+    gcs:send_text(0, "ahrs-source.lua: set SCR_USER1 to rangefinder threshold")
+    return update, 1000
+  end
+
+  -- check GPS speed accuracy threshold has been set
+  local gps_speedaccuracy_thresh = param:get('SCR_USER2')  -- SCR_USER2 holds GPS speed accuracy threshold
+  if (gps_speedaccuracy_thresh == nil) or (gps_speedaccuracy_thresh <= 0) then
+    gcs:send_text(0, "ahrs-source.lua: set SCR_USER2 to GPS speed accuracy threshold")
+    return update, 1000
+  end
+
+  -- check external nav innovation threshold has been set
+  local extnav_innov_thresh = param:get('SCR_USER3')  -- SCR_USER3 holds Non-GPS vertical velocity innovation
+  if (extnav_innov_thresh == nil) or (extnav_innov_thresh <= 0) then
+    gcs:send_text(0, "ahrs-source.lua: set SCR_USER3 to ExtNav innovation threshold")
+    return update, 1000
+  end
+
+  -- check if GPS speed accuracy is over threshold
+  local gps_speed_accuracy = gps:speed_accuracy(0)
+  local gps_over_threshold = (gps_speed_accuracy == nil) or (gps:speed_accuracy(0) > gps_speedaccuracy_thresh)
+  local gps_usable = (gps_speed_accuracy ~= nil) and (gps_speed_accuracy <= gps_usable_accuracy)
+
+  -- get external nav innovations from ahrs
+  local extnav_innov = Vector3f()
+  local extnav_var = Vector3f()
+  extnav_innov, extnav_var = ahrs:get_vel_innovations_and_variances_by_source(6)
```

Comment:
the "6" is not an instance number but rather a source (3=GPS, 4=Beacon, 5=OpticalFlow, 6=ExtNav).  What we need is the innovations if we were to try and fuse external nav into the velocity estimate.  It would be nice if we had an enum of the possible sources but I ran into troubles exposing the enum to scripting